File: library/src/main/java/com/xwray/groupie/NestedGroup.java
package com.xwray.groupie;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * A base implementation of the Group interface, which supports nesting of Groups to arbitrary depth.
 * You can make a NestedGroup which contains only Items, one which contains Groups, or a mixture.
 * <p>
 * It provides support for notifying the adapter about changes which happen in its child groups.
 */
public abstract class NestedGroup implements Group, GroupDataObserver {

    private final GroupDataObservable observable = new GroupDataObservable();

    public int getItemCount() {
        int size = 0;
        for (int i = 0; i < getGroupCount(); i++) {
            Group group = getGroup(i);
            size += group.getItemCount();
        }
        return size;
    }

    protected int getItemCountBeforeGroup(@NonNull final Group group) {
        final int groupIndex = getPosition(group);
        return getItemCountBeforeGroup(groupIndex);
    }

    protected int getItemCountBeforeGroup(final int groupIndex) {
        int size = 0;
        for (int i = 0; i < groupIndex; i++) {
            final Group currentGroup = getGroup(i);
            size += currentGroup.getItemCount();
        }
        return size;
    }

    @NonNull
    public abstract Group getGroup(int position);

    public abstract int getGroupCount();

    @NonNull
    public Item getItem(int position) {
        int previousPosition = 0;

        for (int i = 0; i < getGroupCount(); i++) {
            Group group = getGroup(i);
            int size = group.getItemCount();
            if (size + previousPosition > position) {
                return group.getItem(position - previousPosition);
            }
            previousPosition += size;
        }

        throw new IndexOutOfBoundsException("Wanted item at " + position + " but there are only "
                + getItemCount() + " items");
    }

    public final int getPosition(@NonNull Item item) {
        int previousPosition = 0;

        for (int i = 0; i < getGroupCount(); i++) {
            Group group = getGroup(i);
            int position = group.getPosition(item);
            if (position >= 0) {
                return position + previousPosition;
            }
            previousPosition += group.getItemCount();
        }

        return -1;
    }

    public abstract int getPosition(@NonNull Group group);

    @Override
    public final void registerGroupDataObserver(@NonNull GroupDataObserver groupDataObserver) {
        observable.registerObserver(groupDataObserver);
    }

    @Override
    public void unregisterGroupDataObserver(@NonNull GroupDataObserver groupDataObserver) {
        observable.unregisterObserver(groupDataObserver);
    }

    @CallSuper
    public void add(@NonNull Group group) {
        group.registerGroupDataObserver(this);
    }

    @CallSuper
    public void addAll(@NonNull Collection<? extends Group> groups) {
        for (Group group : groups) {
            group.registerGroupDataObserver(this);
        }
    }

    @CallSuper
    public void add(int position, @NonNull Group group) {
        group.registerGroupDataObserver(this);
    }

    @CallSuper
    public void addAll(int position, @NonNull Collection<? extends Group> groups) {
        for (Group group : groups) {
            group.registerGroupDataObserver(this);
        }
    }

    @CallSuper
    public void remove(@NonNull Group group) {
        group.unregisterGroupDataObserver(this);
    }

    @CallSuper
    public void removeAll(@NonNull Collection<? extends Group> groups) {
        for (Group group : groups) {
            group.unregisterGroupDataObserver(this);
        }
    }

    @CallSuper
    public void replaceAll(@NonNull Collection<? extends Group> groups) {
        final int groupCount = getGroupCount();

        for (int i = groupCount - 1; i >= 0; i--) {
            getGroup(i).unregisterGroupDataObserver(this);
        }

        for (Group group: groups) {
            group.registerGroupDataObserver(this);
        }
    }

    /**
     * Every item in the group still exists but the data in each has changed (e.g. should rebind).
     *
     * @param group
     */
    @CallSuper
    @Override
    public void onChanged(@NonNull Group group) {
        observable.onItemRangeChanged(this, getItemCountBeforeGroup(group), group.getItemCount());
    }

    @CallSuper
    @Override
    public void onItemInserted(@NonNull Group group, int position) {
        observable.onItemInserted(this, getItemCountBeforeGroup(group) + position);
    }

    @CallSuper
    @Override
    public void onItemChanged(@NonNull Group group, int position) {
        observable.onItemChanged(this, getItemCountBeforeGroup(group) + position);
    }

    @CallSuper
    @Override
    public void onItemChanged(@NonNull Group group, int position, Object payload) {
        observable.onItemChanged(this, getItemCountBeforeGroup(group) + position, payload);
    }

    @CallSuper
    @Override
    public void onItemRemoved(@NonNull Group group, int position) {
        observable.onItemRemoved(this, getItemCountBeforeGroup(group) + position);
    }

    @CallSuper
    @Override
    public void onItemRangeChanged(@NonNull Group group, int positionStart, int itemCount) {
        observable.onItemRangeChanged(this, getItemCountBeforeGroup(group) + positionStart, itemCount);
    }

    @CallSuper
    @Override
    public void onItemRangeChanged(@NonNull Group group, int positionStart, int itemCount, Object payload) {
        observable.onItemRangeChanged(this, getItemCountBeforeGroup(group) + positionStart, itemCount, payload);
    }

    @CallSuper
    @Override
    public void onItemRangeInserted(@NonNull Group group, int positionStart, int itemCount) {
        observable.onItemRangeInserted(this, getItemCountBeforeGroup(group) + positionStart, itemCount);
    }

    @CallSuper
    @Override
    public void onItemRangeRemoved(@NonNull Group group, int positionStart, int itemCount) {
        observable.onItemRangeRemoved(this, getItemCountBeforeGroup(group) + positionStart, itemCount);
    }

    @CallSuper
    @Override
    public void onItemMoved(@NonNull Group group, int fromPosition, int toPosition) {
        int groupPosition = getItemCountBeforeGroup(group);
        observable.onItemMoved(this, groupPosition + fromPosition, groupPosition + toPosition);
    }

    @CallSuper
    @Override
    public void onDataSetInvalidated() {
        observable.onDataSetInvalidated();
    }

    /**
     * A group should use this to notify that there is a change in itself.
     *
     * @param positionStart
     * @param itemCount
     */
    @CallSuper
    public void notifyItemRangeInserted(int positionStart, int itemCount) {
        observable.onItemRangeInserted(this, positionStart, itemCount);
    }

    @CallSuper
    public void notifyItemRangeRemoved(int positionStart, int itemCount) {
        observable.onItemRangeRemoved(this, positionStart, itemCount);
    }

    @CallSuper
    public void notifyItemMoved(int fromPosition, int toPosition) {
        observable.onItemMoved(this, fromPosition, toPosition);
    }

    @CallSuper
    public void notifyChanged() {
        observable.onChanged(this);
    }

    @CallSuper
    public void notifyItemInserted(int position) {
        observable.onItemInserted(this, position);
    }

    @CallSuper
    public void notifyItemChanged(int position) {
        observable.onItemChanged(this, position);
    }

    @CallSuper
    public void notifyItemChanged(int position, @Nullable Object payload) {
        observable.onItemChanged(this, position, payload);
    }

    @CallSuper
    public void notifyItemRemoved(int position) {
        observable.onItemRemoved(this, position);
    }

    @CallSuper
    public void notifyItemRangeChanged(int positionStart, int itemCount) {
        observable.onItemRangeChanged(this, positionStart, itemCount);
    }

    @CallSuper
    public void notifyItemRangeChanged(int positionStart, int itemCount, Object payload) {
        observable.onItemRangeChanged(this, positionStart, itemCount, payload);
    }

    @CallSuper
    public void notifyDataSetInvalidated() {
        observable.onDataSetInvalidated();
    }

    /**
     * Iterate in reverse order in case any observer decides to remove themself from the list
     * in their callback
     */
    private static class GroupDataObservable {
        final List<GroupDataObserver> observers = new ArrayList<>();

        void onItemRangeChanged(Group group, int positionStart, int itemCount) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemRangeChanged(group, positionStart, itemCount);
            }
        }

        void onItemRangeChanged(Group group, int positionStart, int itemCount, Object payload) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemRangeChanged(group, positionStart, itemCount, payload);
            }
        }

        void onItemInserted(Group group, int position) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemInserted(group, position);
            }
        }

        void onItemChanged(Group group, int position) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemChanged(group, position);
            }
        }

        void onItemChanged(Group group, int position, Object payload) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemChanged(group, position, payload);
            }
        }

        void onItemRemoved(Group group, int position) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemRemoved(group, position);
            }
        }

        void onItemRangeInserted(Group group, int positionStart, int itemCount) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemRangeInserted(group, positionStart, itemCount);
            }
        }

        void onItemRangeRemoved(Group group, int positionStart, int itemCount) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemRangeRemoved(group, positionStart, itemCount);
            }
        }

        void onItemMoved(Group group, int fromPosition, int toPosition) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onItemMoved(group, fromPosition, toPosition);
            }
        }

        void onChanged(Group group) {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onChanged(group);
            }
        }

        void registerObserver(GroupDataObserver observer) {
            synchronized(observers) {
                if (observers.contains(observer)) {
                    throw new IllegalStateException("Observer " + observer + " is already registered.");
                }
                observers.add(observer);
            }
        }

        void unregisterObserver(GroupDataObserver observer) {
            synchronized(observers) {
                int index = observers.indexOf(observer);
                observers.remove(index);
            }
        }

        void onDataSetInvalidated() {
            for (int i = observers.size() - 1; i >= 0; i--) {
                observers.get(i).onDataSetInvalidated();
            }
        }
    }
}

File: library-kotlin/src/main/java/com/wray/groupiekotlin/NestedGroup.kt
package com.wray.groupiekotlin

import androidx.annotation.CallSuper

abstract class NestedGroup: Group, GroupDataObserver {

    private val observable = GroupDataObservable()

    abstract fun getGroup(position: Int): Group

    abstract val groupCount: Int

    abstract fun getPosition(group: Group): Int

    protected fun getItemCountBeforeGroup(group: Group): Int {
        val groupIndex = getPosition(group)
        return getItemCountBeforeGroupIndex(groupIndex)
    }

    protected fun getItemCountBeforeGroupIndex(groupIndex: Int): Int {
        var size = 0
        for (i in 0 until groupIndex) {
            val currentGroup = getGroup(i)
            size += currentGroup.itemCount
        }
        return size
    }

    @CallSuper
    open fun add(group: Group) {
        group.registerGroupDataObserver(this)
    }

    @CallSuper
    open fun <T : Group> addAll(groups: Collection<T>) {
        groups.forEach { group -> group.registerGroupDataObserver(this) }
    }

    @CallSuper
    open fun add(position: Int, group: Group) {
        add(group)
    }

    @CallSuper
    open fun <T : Group> addAll(position: Int, groups: Collection<T>) {
        addAll(groups)
    }

    @CallSuper
    open fun remove(group: Group) {
        group.unregisterGroupDataObserver(this)
    }

    @CallSuper
    open fun <T : Group> removeAll(groups: Collection<T>) {
        groups.forEach { group -> group.unregisterGroupDataObserver(this) }
    }

    /**
     * A group should use this to notify that there is a change in itself.
     *
     * @param positionStart
     * @param itemCount
     */
    @CallSuper
    open fun notifyItemRangeInserted(positionStart: Int, itemCount: Int) {
        observable.onItemRangeInserted(this, positionStart, itemCount)
    }

    @CallSuper
    open fun notifyItemRangeRemoved(positionStart: Int, itemCount: Int) {
        observable.onItemRangeRemoved(this, positionStart, itemCount)
    }

    @CallSuper
    open fun notifyItemMoved(fromPosition: Int, toPosition: Int) {
        observable.onItemMoved(this, fromPosition, toPosition)
    }

    @CallSuper
    open fun notifyChanged() {
        observable.onChanged(this)
    }

    @CallSuper
    open fun notifyItemInserted(position: Int) {
        observable.onItemInserted(this, position)
    }

    @CallSuper
    open fun notifyItemChanged(position: Int) {
        observable.onItemChanged(this, position)
    }

    @CallSuper
    open fun notifyItemChanged(position: Int, payload: Any) {
        observable.onItemChanged(this, position, payload)
    }

    @CallSuper
    open fun notifyItemRemoved(position: Int) {
        observable.onItemRemoved(this, position)
    }

    @CallSuper
    open fun notifyItemRangeChanged(positionStart: Int, itemCount: Int) {
        observable.onItemRangeChanged(this, positionStart, itemCount)
    }

    @CallSuper
    open fun notifyItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any) {
        observable.onItemRangeChanged(this, positionStart, itemCount, payload)
    }

    //region: Group interface

    final override val itemCount: Int
        get() {
            var size = 0
            for (i in 0 until groupCount) {
                val group = getGroup(i)
                size += group.itemCount
            }
            return size
        }

    final override fun getItem(position: Int): AnyItem {
        var previousPosition = 0

        for (i in 0 until groupCount) {
            val group = getGroup(i)
            val size = group.itemCount
            if (size + previousPosition > position) {
                return group.getItem(position - previousPosition)
            }
            previousPosition += size
        }

        throw IndexOutOfBoundsException(
            "Wanted item at " + position + " but there are only "
                + itemCount + " items"
        )
    }

    final override fun getPosition(item: AnyItem): Int {
        var previousPosition = 0

        for (i in 0 until groupCount) {
            val group = getGroup(i)
            val position = group.getPosition(item)
            if (position >= 0) {
                return position + previousPosition
            }
            previousPosition += group.itemCount
        }

        return -1
    }

    final override fun registerGroupDataObserver(groupDataObserver: GroupDataObserver) {
        observable.registerObserver(groupDataObserver)
    }

    final override fun unregisterGroupDataObserver(groupDataObserver: GroupDataObserver) {
        observable.unregisterObserver(groupDataObserver)
    }

    //endregion: Group interface

    //region: GroupDataObserver interface


    /**
     * Every item in the group still exists but the data in each has changed (e.g. should rebind).
     *
     * @param group
     */
    @CallSuper
    override fun onChanged(group: Group) {
        observable.onItemRangeChanged(this, getItemCountBeforeGroup(group), group.itemCount)
    }

    @CallSuper
    override fun onItemInserted(group: Group, position: Int) {
        observable.onItemInserted(this, getItemCountBeforeGroup(group) + position)

    }

    @CallSuper
    override fun onItemChanged(group: Group, position: Int) {
        observable.onItemChanged(this, getItemCountBeforeGroup(group) + position)
    }

    @CallSuper
    override fun onItemChanged(group: Group, position: Int, payload: Any) {
        observable.onItemChanged(this, getItemCountBeforeGroup(group) + position, payload)
    }

    @CallSuper
    override fun onItemRemoved(group: Group, position: Int) {
        observable.onItemRemoved(this, getItemCountBeforeGroup(group) + position)
    }

    @CallSuper
    override fun onItemRangeChanged(group: Group, positionStart: Int, itemCount: Int) {
        observable.onItemRangeChanged(this, getItemCountBeforeGroup(group) + positionStart, itemCount)
    }

    @CallSuper
    override fun onItemRangeChanged(group: Group, positionStart: Int, itemCount: Int, payload: Any) {
        observable.onItemRangeChanged(this, getItemCountBeforeGroup(group) + positionStart, itemCount, payload)
    }

    @CallSuper
    override fun onItemRangeInserted(group: Group, positionStart: Int, itemCount: Int) {
        observable.onItemRangeInserted(this, getItemCountBeforeGroup(group) + positionStart, itemCount)
    }

    @CallSuper
    override fun onItemRangeRemoved(group: Group, positionStart: Int, itemCount: Int) {
        observable.onItemRangeRemoved(this, getItemCountBeforeGroup(group) + positionStart, itemCount)
    }

    @CallSuper
    override fun onItemMoved(group: Group, fromPosition: Int, toPosition: Int) {
        val groupPosition = getItemCountBeforeGroup(group)
        observable.onItemMoved(this, groupPosition + fromPosition, groupPosition + toPosition)
    }

    //endregion: GroupDataObserver interface

}

/**
 * Iterate using a listIterator to support items that might remove themselves while we iterate
 */
private class GroupDataObservable {

    private val observers = mutableListOf<GroupDataObserver>()

    internal fun onItemRangeChanged(group: Group, positionStart: Int, itemCount: Int) {
        observers.listIterator().forEach { observer ->
            observer.onItemRangeChanged(group, positionStart, itemCount)
        }
    }

    internal fun onItemRangeChanged(group: Group, positionStart: Int, itemCount: Int, payload: Any) {
        observers.listIterator().forEach { observer ->
            observer.onItemRangeChanged(group, positionStart, itemCount, payload)
        }
    }

    internal fun onItemInserted(group: Group, position: Int) {
        observers.listIterator().forEach { observer ->
            observer.onItemInserted(group, position)
        }
    }

    internal fun onItemChanged(group: Group, position: Int) {
        observers.listIterator().forEach { observer ->
            observer.onItemChanged(group, position)
        }
    }

    internal fun onItemChanged(group: Group, position: Int, payload: Any) {
        observers.listIterator().forEach { observer ->
            observer.onItemChanged(group, position, payload)
        }
    }

    internal fun onItemRemoved(group: Group, position: Int) {
        observers.listIterator().forEach { observer ->
            observer.onItemRemoved(group, position)
        }
    }

    internal fun onItemRangeInserted(group: Group, positionStart: Int, itemCount: Int) {
        observers.listIterator().forEach { observer ->
            observer.onItemRangeInserted(group, positionStart, itemCount)
        }
    }

    internal fun onItemRangeRemoved(group: Group, positionStart: Int, itemCount: Int) {
        observers.listIterator().forEach { observer ->
            observer.onItemRangeRemoved(group, positionStart, itemCount)
        }
    }

    internal fun onItemMoved(group: Group, fromPosition: Int, toPosition: Int) {
        observers.listIterator().forEach { observer ->
            observer.onItemMoved(group, fromPosition, toPosition)
        }
    }

    internal fun onChanged(group: Group) {
        observers.listIterator().forEach { observer ->
            observer.onChanged(group)
        }
    }

    internal fun registerObserver(observer: GroupDataObserver) {
        synchronized(observers) {
            check(!observers.contains(observer)) { "Observer $observer is already registered." }
            observers.add(observer)
        }
    }

    internal fun unregisterObserver(observer: GroupDataObserver) {
        synchronized(observers) {
            observers.remove(observer)
        }
    }
}